### PR TITLE
Fix sidecar container exec (#1216)

### PR DESCRIPTION
**Pull request by rmodpur**
### Checklist
- [x] The title of this PR would make a good line in Acorn's Release Note's Changelog
- [x] The title of this PR ends with a link to the main issue being address in paranthesis, like: `This is a title (#1216)`. [Here's an example](https://github.com/acorn-io/acorn/pull/1199)
- [x] All relevant issues are referenced in the PR description. *NOTE: don't use [GitHub keyworkds](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) that auto-close issues*
- [x] Commits follow [contributing guidance](https://github.com/acorn-io/acorn/blob/main/CONTRIBUTING.md#commits)
- [x] Automated tests added to cover the changes. If tests couldn't be added, an explanation is provided in the Verification and Testing section
- [x] Changes to user-facing functionality, API, CLI, and upgrade impacts are clearly called out in PR description

### Description
Suppose there is an app named `app` with one container named `web` that has a sidecar named `sidecar`

Before the fix these worked:
a. `acorn exec -c web app` (Note that app is not entered twice)
b. `acorn exec app.web-5c979bdb7b-vzq7j` (the extra characters come from the actual pod running in Kubernetes)

With the fix the below mentioned variations work as well:
c. `acorn exec app.web-5c979bdb7b-vzq7j.web`
d. `acorn exec -c web.sidecar app`
e. `acorn exec app.web-5c979bdb7b-vzq7j.sidecar`

The -c flag accepts the container name as defined in the Acornfile and not the Kubernetes pod name.

For #1216

### CODE REVIEW

**Comment by rmodpur** [comment]
@cjellick could you please review this ?

File: pkg/server/registry/apigroups/acorn/containers/translator.go
Comment (jacobdonenfeld) [inline pkg/server/registry/apigroups/acorn/containers/translator.go:34]: Hesitant on this function, especially without seeing any unit tests covering the different ways the for loop can execute. Other then this, lgtm. Nice to see tests written and all passing!

Comment (rmodpur) [reply pkg/server/registry/apigroups/acorn/containers/translator.go:34]: I have added the unit tests

File: pkg/server/registry/apigroups/acorn/containers/translator.go
Comment (jacobdonenfeld) [inline pkg/server/registry/apigroups/acorn/containers/translator.go:38]: NIT: can refactor the for loop to 
```	// stop when name = <pod name> + "." + <container name>
	for strings.Count(name, ".") > 1 {
		prefix, suffix, _ := strings.Cut(name, ".")
		app := &v1.AppInstance{}
		err := t.client.Get(ctx, router.Key(namespace, prefix), app)
		if err != nil {
			return namespace, name, err
		}

		name = suffix
		namespace = app.Status.Namespace
	}
```

**Review by jacobdonenfeld — APPROVED** [review]
LGTM, one small nit. Thanks for adding tests! :)

File: pkg/server/registry/apigroups/acorn/containers/translator.go
Comment (rmodpur) [reply pkg/server/registry/apigroups/acorn/containers/translator.go]: Done
Thanks for the review !

**Review by tylerslaton — COMMENTED** [review]
Tested these changes out locally and they seem to work great! 

Only thing I want to highlight is that `acorn exec -c` doesn't seem to work nicely with sidecars right now so I have to always manually select the sidecar from the interactive prompt.

**Comment by rmodpur** [comment]
> Tested these changes out locally and they seem to work great!
> 
> Only thing I want to highlight is that `acorn exec -c` doesn't seem to work nicely with sidecars right now so I have to always manually select the sidecar from the interactive prompt.

What are you using as container name for `-c` ? the entire container public name or just the last part of it?
I tried `-c` with entire name(`<app-name>.<pod-name>.<container-name>`) it worked for both main & sidecar container.

**Comment by tylerslaton** [comment]
Interesting, maybe there is some environment skew on my end? I'm seeing the following:

<details>

<summary>This works like I'd expect</summary>

  ```bash
  $ acorn containers bold-leaf            
  NAME                                     APP         IMAGE                                          STATE     RESTARTCOUNT   CREATED   MESSAGE
  bold-leaf.app-65dffdbd9c-9f92g.app       bold-leaf   public.ecr.aws/docker/library/nginx:latest     running   0              61m ago   
  bold-leaf.app-65dffdbd9c-9f92g.sidecar   bold-leaf   public.ecr.aws/docker/library/busybox:latest   running   0              61m ago    
  $ acorn exec bold-leaf.app-65dffdbd9c-9f92g.sidecar
  / #
  ```

</details>

<details>

<summary>This fails even though it seems that it shouldn't. Am I doing something wrong here?</summary>

```bash
# same containers seen in the first drop down
$ acorn exec -c bold-leaf.app-65dffdbd9c-9f92g.sidecar bold-leaf
  ✗  ERROR:  failed to find any containers for app bold-leaf
```

</details>

<details> 

<summary>As an aside, it would a nice UX if this worked for sidecars since it works for regular containers.</summary>

```bash
$ acorn exec -c app.sidecar bold-leaf
✗  ERROR:  failed to find any containers for app bold-leaf
$ acorn exec -c app bold-leaf
root@app:/#
```

</details>

**Comment by rmodpur** [comment]
Your 1st & 2nd attempts should work, 3rd one should not work at all(for both sidecars & main container) with these changes
There were some changes in api server code as well, could you check docker image used for api server ?

**Review by tylerslaton — APPROVED** [review]
Looks like it must have been environment skew for me because the 2nd attempt started working. This looks good to me now, thank you for the contribution.

**Comment by thedadams** [comment]
@rmodpur Are you saying that `acorn exec -c app bold-leaf` will no longer work with these changes?

**Comment by rmodpur** [comment]
> @rmodpur Are you saying that `acorn exec -c app bold-leaf` should no longer work with these changes?

Yes, the assumption is the end user will only interact with acorn objects using it's public name which for the container is it's full name

**Comment by thedadams** [comment]
That command works currently and should continue to work after these changes. Additionally, `acorn exec -c app.sidecar bold-leaf` should work, too, but we won't block merging this for this additional enhancement.

For now, I will dismiss the approvals so this change doesn't get merged accidentally until it is ensured that this command continues to work.

**Comment by rmodpur** [comment]
Let's assume there are two containers for an app named `app`
1. `app.pod.main`
2. `app.pod.sidecar`

What are the variants that should work ?
a. `acorn exec app.pod.sidecar` (works now)
b. `acorn exec -c app.pod.sidecar app` (works now)
c. `acorn exec -c sidecar app` (doesn't work)

I understand that this PR may not be right place for it but I do have one suggestion, I think `-c` flag is redundant because the way exec is defined `exec [flags] APP_NAME|CONTAINER_NAME CMD` an user should be able to exec using these ways
1. `acorn exec` (will be asked for app first & then container)
2. `acorn exec app` (will be asked for a choice of container if more than one)
3.  `acorn exec app.pod.sidecar`

**Comment by thedadams** [comment]
I think there is some confusion here on the what I am talking about. Let's suppose I have an app named `app` with one container named `web` that has a sidecar named `sidecar`.

Currently, in `main`, the following work:
a. `acorn exec -c web app` (Note that `app` is not entered twice)
b. `acorn exec app.web-5c979bdb7b-vzq7j` (the extra characters come from the actual pod running in Kubernetes)

These should continue to work with your changes.

The following do not work in `main` currently:
a. `acorn exec -c web-5c979bdb7b-vzq7j app`
b. `acorn exec app.web`
c. `acorn exec app.web-5c979bdb7b-vzq7j.web`
d. `acorn exec -c web.sidecar app`
e. `acorn exec app.web-5c979bdb7b-vzq7j.sidecar`

As I understand it, this change was to ensure e. works (and hopefully d. too).

I understand what you are saying about the `-c` flag feeling redundant. However, it is there for automated (CI/GitOps) interactions with Acorn where being given a selection to chose from doesn't work. Additionally, the `-c` flag accepts the container name as defined in the Acornfile and not the Kubernetes pod name. Ideally, the Kubernetes-ness is abstracted from the user as much as possible.

**Comment by rmodpur** [comment]
@thedadams could you please review

Apart from the currently working ones you mentioned above these work with the fix
c. `acorn exec app.web-5c979bdb7b-vzq7j.web`
d. `acorn exec -c web.sidecar app`
e. `acorn exec app.web-5c979bdb7b-vzq7j.sidecar`

File: pkg/server/registry/apigroups/acorn/containers/translator_test.go
Comment (rmodpur) [reply pkg/server/registry/apigroups/acorn/containers/translator_test.go:83]: container's `FromPublicName` method is never called with `container.sidecar` pattern, in this case the cli code for `exec` lists all the containers for the `app` & matches  `container.sidecar` with `app.container-5c979bdb7b-vzq7j.sidecar` for example.

**Review by tylerslaton — APPROVED** [review]

**Review by thedadams — APPROVED** [review]

**Pull request merged**
